scripts/circle-env.js

```diff
@@ -0,0 +1,58 @@
+/* eslint-disable no-console */
+const fs = require('fs').promises
+
+async function loadInternalTaskData () {
+  const filename = process.env.CIRCLE_INTERNAL_CONFIG
+
+  if (!filename) throw new Error('Missing CIRCLE_INTERNAL_CONFIG environment variable, cannot load Circle task data.')
+
+  const taskDataJson = await fs.readFile(filename, 'utf8')
+
+  try {
+    return JSON.parse(taskDataJson)
+  } catch (err) {
+    throw new Error(`An error occurred while parsing the Circle task data: ${err}`)
+  }
+}
+
+// check if the project env canary and context canary are both present to verify that this script is reading the right env
+async function checkCanaries () {
+  if (!process.env.CI) console.warn('This script will not work outside of CI.')
+
+  const circleEnv = await readCircleEnv()
+
+  if (!circleEnv.MAIN_CANARY) throw new Error('Missing MAIN_CANARY.')
+
+  if (!circleEnv.CONTEXT_CANARY) throw new Error('Missing CONTEXT_CANARY. Does this job have the test-runner:env-canary context?')
+}
+
+// Returns a map of environment variables defined for this job. `readCircleEnv()` differs from `process.env` - it will
+// only return environment variables explicitly specified for this job by CircleCI project env and contexts
+// NOTE: this Circle API is not stable, and yet it is the only way to access this information.
+async function readCircleEnv () {
+  const taskData = await loadInternalTaskData()
+
+  try {
+    const circleEnv = taskData['Dispatched']['TaskInfo']['Environment']
```

Comment:
Is this API documented/open source anywhere? Could be useful to have a link to refer to if something gets updated?

Reply:
No, it's an internal API. Do you think a comment would be useful?

```suggestion
    // if this starts failing, try SSHing into a CircleCI job and see what changed in the $CIRCLE_INTERNAL_CONFIG file's schema
    const circleEnv = taskData['Dispatched']['TaskInfo']['Environment']
```